File: 백준/Gold/16398. 행성 연결/행성 연결.cc
// Authored by : chjh2129
#include <bits/stdc++.h>

using namespace std;

using ti = tuple<int,int,int>;

int n;
vector<ti> edge; // 간선 정보
vector<int> p(1005, -1); // 부모 정보

int find(int x){ return (p[x] < 0 ? x : p[x] = find(p[x])); }

bool merge(int x, int y){
  x = find(x);
  y = find(y);
  if(x == y) return 0; // 유니온 실패
  if(p[x] == p[y]) --p[x];
  if(p[x] > p[y]) swap(x, y);
  p[y] = x;
  return 1; // 유니온 성공
}

int main(){
  cin.tie(nullptr)->sync_with_stdio(false);

  // input
  cin >> n;
  for(int i = 1; i <= n; i++){
    int cost;
    for(int j = 1; j <= n; j++){
      cin >> cost;
      if(i >= j) continue;
      edge.push_back({cost, i, j});
    }
  }

  // solve
  int cnt{};
  long long ans{};
  sort(edge.begin(), edge.end());
  for(auto &[cost, u, v] : edge){
    if(!merge(u, v)) continue; // 유니온 실패 시 넘어감
    cnt++;
    ans += cost;
    if(cnt == n-1) break; // 간선을 n-1개 선택했으면 루프를 종료
  }

  // output
  cout << ans;
}

File: 백준/Gold/16398. 행성 연결/행성 연결.py
#Authored by : chjh2129
import sys

input = sys.stdin.readline

def find(x: int, p: list[int]) -> int:
  root:int = x
  while p[root] > 0:
    root = p[root]
  
  while p[x] > 0:
    nxt, p[x] = p[x], root
    x = nxt
  
  return root

def union(x: int, y: int, p: list[int]) -> bool:
  x = find(x, p)
  y = find(y, p)
  if x == y: return False
  if p[x] == p[y]: p[x] -= 1
  if p[x] > p[y]: x, y = y, x
  p[y] = x
  return True

def main() -> None:
  # input & init
  N: int = int(input())
  P: list[int] = [ -1 for _ in range(N+1) ]
  Edges: list[tuple] = []

  for i in range(1, N+1):
    cost: list[int] = list(map(int, input().split()))
    for j in range(i+1, N+1):
      Edges.append((cost[j-1], i, j))

  # solve
  Edges.sort(key=lambda x : x[0])

  cnt, ans = 0, 0
  for cost, u, v in Edges:
    if union(u, v, P):
      cnt += 1
      ans += cost
      if cnt == N - 1: break

  # output
  print(ans)
sys.exit(main())